src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,296 @@
+#include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/ai/intent/intent.h"
+#include "software/ai/intent/move_intent.h"
+#include "software/ai/passing/pass.h"
+#include "software/geom/algorithms/closest_point.h"
+
+struct ReceiverFSM
+{
+    class OneTouchShotState;
+    class ReceiveAndDribbleState;
+    class UndecidedState;
+
+    struct ControlParams
+    {
+        // The pass to receive
+        std::optional<Pass> pass = std::nullopt;
+    };
+
+    DEFINE_UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS
+
+    // The minimum proportion of open net we're shooting on vs the entire size of the net
+    // that we require before attempting a shot
+    static constexpr double MIN_SHOT_NET_PERCENT_OPEN    = 0.3;
+    static constexpr double MAX_SPEED_FOR_ONE_TOUCH_SHOT = 6.5;
+    static constexpr double MIN_PASS_START_SPEED         = 1.0;
+
+    // The maximum deflection angle that we will attempt a one-touch kick towards the
+    // enemy goal with
+    static constexpr Angle MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT = Angle::fromDegrees(90);
+
+    /**
+     * Given a shot and the ball, figures out the angle the robot should be facing
+     * to perform a one-touch shot.
+     *
+     * @param shot The shot to take
+     * @param ball The ball on the field
+     */
+    static Angle getOneTimeShotDirection(const Ray& shot, const Ball& ball)
+    {
+        Vector shot_vector = shot.toUnitVector();
+        Angle shot_dir     = shot.getDirection();
+
+        Vector ball_vel    = ball.velocity();
+        Vector lateral_vel = ball_vel.project(shot_vector.perpendicular());
+
+        // The lateral speed is roughly a measure of the lateral velocity we need to
+        // "cancel out" in order for our shot to go in the expected direction.
+        // The scaling factor of 0.3 is a magic number that was carried over from the old
+        // code. It seems to work well on the field.
+        double lateral_speed = 0.3 * lateral_vel.length();
+
+        // This kick speed is based off of the value used in the firmware `MovePrimitive`
+        // when autokick is enabled
+        double kick_speed = BALL_MAX_SPEED_METERS_PER_SECOND - 1;
+        Angle shot_offset = Angle::asin(lateral_speed / kick_speed);
+
+        // check which direction the ball is going in so we can decide which direction to
+        // apply the offset in
+        if (lateral_vel.dot(shot_vector.rotate(Angle::quarter())) > 0)
+        {
+            // need to go clockwise
+            shot_offset = -shot_offset;
+        }
+        return shot_dir + shot_offset;
+    }
+
+    /**
+     * Figures out the location of the one-time shot and orientation the robot should face
+     *
+     * @param robot The robot performing the one-touch
+     * @param ball The ball on the field
+     * @param best_shot_target The point to shoot at
+     */
+    static Shot getOneTimeShotPositionAndOrientation(const Robot& robot, const Ball& ball,
```

Comment:
nit: also here, can this be called getOneTouchShotPositionAndOrientation?

Reply:
yup updated!